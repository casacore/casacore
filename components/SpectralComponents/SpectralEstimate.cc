//# SpectralEstimate.cc: Get an initial estimate for spectral lines
//# Copyright (C) 2001,2002,2003,2004
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

//# Includes
#include <components/SpectralComponents/SpectralEstimate.h>
#include <casa/Arrays/Vector.h>
#include <casa/BasicMath/Math.h>

namespace casa { //# NAMESPACE CASA - BEGIN

//# Constructors
SpectralEstimate::SpectralEstimate(const uInt maxpar) :
  useWindow_p(False), rms_p(0), cutoff_p(0),
  windowLow_p(0), windowEnd_p(0),
  regionLow_p(0), regionEnd_p(0),
  q_p(2), sigmin_p(0),
  deriv_p(0), slist_p(maxpar), lprof_p(0) {
  setQ();
}

SpectralEstimate::SpectralEstimate(const Double rms,
				   const Double cutoff, const Double minsigma,
				   const uInt maxpar) :
  useWindow_p(False), rms_p(rms), cutoff_p(cutoff),
  windowLow_p(0), windowEnd_p(0),
  regionLow_p(0), regionEnd_p(0),
  q_p(2), sigmin_p(minsigma),
  deriv_p(0), slist_p(maxpar), lprof_p(0) {
  setQ();
}


SpectralEstimate::SpectralEstimate(const SpectralEstimate &other) :
  useWindow_p(other.useWindow_p), rms_p(other.rms_p), cutoff_p(other.cutoff_p),
  windowLow_p(other.windowLow_p), windowEnd_p(other.windowEnd_p),
  regionLow_p(other.regionLow_p), regionEnd_p(other.regionEnd_p),
  q_p(other.q_p), sigmin_p(other.sigmin_p),
  deriv_p(0), slist_p(other.slist_p), lprof_p(other.lprof_p) {
  setQ(q_p);
  deriv_p = new Double[lprof_p];
  for (uInt i=0; i<lprof_p; i++) deriv_p[i] = other.deriv_p[i];
}

SpectralEstimate::~SpectralEstimate() {
  delete [] deriv_p; deriv_p = 0; lprof_p = 0;
}

SpectralEstimate &SpectralEstimate::operator=(const SpectralEstimate &other) {
  if (this != &other) {
    useWindow_p = other.useWindow_p; 
    rms_p = other.rms_p; 
    cutoff_p = other.cutoff_p;
    windowLow_p = other.windowLow_p;
    windowEnd_p = other.windowEnd_p;
    regionLow_p = other.regionLow_p;
    regionEnd_p = other.regionEnd_p;
    q_p = other.q_p;
    sigmin_p = other.sigmin_p;
    deriv_p = 0;
    slist_p = other.slist_p;
    lprof_p = other.lprof_p;
    setQ(q_p);
    deriv_p = new Double[lprof_p];
    for (uInt i=0; i<lprof_p; i++) deriv_p[i] = other.deriv_p[i];
  };
  return *this;
}

void SpectralEstimate::setRMS(const Double rms) {
  rms_p = abs(rms);
}

void SpectralEstimate::setCutoff(const Double cutoff) {
  cutoff_p = max(0.0, cutoff);
}

void SpectralEstimate::setMinSigma(const Double minsigma) {
  sigmin_p = max(0.0, minsigma);
}

void SpectralEstimate::setQ(const uInt q) {
  q_p = max(1, Int(q));
  a_p = 90.0/(q_p*(q_p+1)*(4*q_p*q_p-1)*(2*q_p+3));
  b_p = (q_p*(q_p+1))/3.0;
}

void SpectralEstimate::setRegion(const Int lo, const Int hi) {
  regionLow_p = regionEnd_p = 0;
  if (hi > lo) {
    regionLow_p = lo;
    regionEnd_p = hi;
  };
}

void SpectralEstimate::setWindowing(const Bool win) {
  useWindow_p = win;
}

void SpectralEstimate::setMaxN(const uInt maxpar) {
  slist_p.set(maxpar);
}

} //# NAMESPACE CASA - END


//# Cater for Double and Float
#ifdef AIPS_NO_TEMPLATE_SRC
#include <components/SpectralComponents/Spectral2Estimate.tcc>

namespace casa { //# NAMESPACE CASA - BEGIN
template SpectralList const & SpectralEstimate::estimate<Float>(Vector<Float> const &, Vector<Float> *);
template SpectralList const & SpectralEstimate::estimate<Float>(Vector<Float> const &, Vector<Float> const &);
template SpectralElement SpectralEstimate::convertElement<Float>(Vector<Float> const &,
                                                                 SpectralElement const &) const;
template void SpectralEstimate::findga<Float>(Vector<Float> const &); 
template uInt SpectralEstimate::window<Float>(Vector<Float> const &);
template void SpectralEstimate::findc2<Float>(Vector<Float> const &); 

template SpectralList const & SpectralEstimate::estimate<Double>(Vector<Double> const &, Vector<Double> *);
template SpectralList const & SpectralEstimate::estimate<Double>(Vector<Double> const &, Vector<Double> const &);
template SpectralElement SpectralEstimate::convertElement<Double>(Vector<Double> const &,
                                                                 SpectralElement const &) const;
template void SpectralEstimate::findga<Double>(Vector<Double> const &);
template uInt SpectralEstimate::window<Double>(Vector<Double> const &);
template void SpectralEstimate::findc2<Double>(Vector<Double> const &);
} //# NAMESPACE CASA - END
#endif
