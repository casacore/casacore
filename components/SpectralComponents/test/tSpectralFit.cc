//# tSpectralFit.cc -- test SpectralElement; SpectralEstimate and SpectralFit
//# Copyright (C) 2001,2002,2004
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This program is free software; you can redistribute it and/or modify it
//# under the terms of the GNU General Public License as published by the Free
//# Software Foundation; either version 2 of the License, or (at your option)
//# any later version.
//#
//# This program is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
//# more details.
//#
//# You should have received a copy of the GNU General Public License along
//# with this program; if not, write to the Free Software Foundation, Inc.,
//# 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

//# Includes
#include <casa/aips.h>
#include <casa/Arrays/MaskArrLogi.h>
#include <casa/Arrays/ArrayMath.h>
#include <casa/Arrays/ArrayIO.h>
#include <casa/Arrays/Vector.h>
#include <casa/BasicMath/Random.h>
#include <casa/Containers/Record.h>
#include <casa/Exceptions/Error.h>
#include <casa/Utilities/Assert.h>
#include <components/SpectralComponents/SpectralEstimate.h>
#include <components/SpectralComponents/SpectralFit.h>
#include <components/SpectralComponents/SpectralList.h>
#include <casa/iostream.h>

#include <casa/namespace.h>

int main()
{
      
  // RecordInterface
  {
    cout << "Test RecordInterface" << endl;
    cout << "---------------------------------------------------" << endl;
    SpectralList list;
    GaussianSpectralElement gEl(1.0, 10.0, 2.0);
    PolynomialSpectralElement pEl(1);
    list.add (gEl);
    list.add(pEl);

    cout << "toRecord" << endl;
    Record rec;    
    list.toRecord(rec);
    cout << "fromRecord" << endl;
    SpectralList list2;
    String errMsg;
    if (!list2.fromRecord(errMsg, rec)) {
       throw(AipsError(errMsg));
    }
    Double tol(1.0e-6);
    AlwaysAssert(list.nelements()==list.nelements(),AipsError);
    for (uInt i=0; i<list.nelements(); i++) {
       AlwaysAssert (list[i]->getType()==list2[i]->getType(), AipsError);
       Vector<Double> p1,p2;
       list[i]->get(p1);
       list2[i]->get(p2);
       AlwaysAssert(allNear(p1,p2,tol),AipsError);
    }
    cout << endl;
  }

  // get estimates
  {
    cout << "Test SpectralEstimate" << endl;
    cout << "---------------------------------------------------" << endl;
    const Int NPAR = 9;
    const Int NPTS = 100;
    uInt q = 2;
    Int r;
    Int n = NPTS;
    Int np = NPAR;
    Double rms = 0.5;
    Double cutoff= 1.0;
    Double minsig = 0.5;
    Double par[NPAR];
    Vector<Float> y(NPTS);
    Vector<Float> xy(NPTS);

    par[0] = 5.0; par[1] = 45.0; par[2] = 1.1;
    par[3] = 4.0; par[4] = 50.0; par[5] = 1.1;
    par[6] = 6.0; par[7] = 55.0; par[8] = 1.1;
    for (Int i = 0; i<n; i++) {
      y(i) = 0.0;
      xy(i) = i;
      for (Int j = 0; j < np; j += 3) {
	if (par[j] != 0.0) {
	  Double a = par[j];
	  Double c = (par[j+1] - Double(i));
	  Double s = par[j+2];
	  y(i) += a * exp( -0.5 * c / s * c / s );
	};
      };
    };
    cout << "Made spectrum with components: " << endl;
    for (uInt i=0; i<3; i++) {
      cout << GaussianSpectralElement(
			      par[3*i+0], par[3*i+1], par[3*i+2]) << endl;
    };
    {
      SpectralEstimate est(rms, cutoff, minsig);
      Vector<Float> deriv(n);
      const SpectralList &slist = est.estimate(y, &deriv);
      r = slist.nelements();
      cout << "Found (using rms, cutoff, minsig, q): (" <<
	rms << ", " << cutoff << ", " << minsig << ", " << q << ") " <<
	r << " estimates" << endl;
      for (Int i=0; i<r; i++) {
	cout << "Estimate " << i << ": " << *slist[i] << endl;
      };
      Vector<Float> res(NPTS);
      res = y;
      slist.residual(res);
      cout << "Minimum, Maximum residuals: " << min(res) << ", " <<
	max(res) << endl;
      for (Int j=0; j<r; j++) {
    	  for (Int i = 0; i<n; i++) {
    		  res(i) = slist[j]->operator()(i);
    	  }
      }
    }
    {
      SpectralEstimate est(rms, cutoff, minsig);
      est.setWindowing(True);
      const SpectralList &slist = est.estimate(xy, y);
      r = slist.nelements();
      cout << "Window (using rms, cutoff, minsig, q): (" <<
	rms << ", " << cutoff << ", " << minsig << ", " << q << ") " <<
	r << " estimates" << endl;
      for (Int i=0; i<r; i++) {
	cout << "Estimate " << i << ": " << *slist[i] << endl;
      };
    }
    {
      q = 1;
      SpectralEstimate est(rms, cutoff, minsig);
      est.setQ(q);
      const SpectralList &slist = est.estimate(xy, y);
      r = slist.nelements();
      cout << "Found (using rms, cutoff, minsig, q): (" <<
	rms << ", " << cutoff << ", " << minsig << ", " << q << ") " <<
	r << " estimates" << endl;
      for (Int i=0; i<r; i++) {
	cout << "Estimate " << i << ": " << *slist[i] << endl;
      };
    }
    {
      q = 1;
      rms = cutoff = minsig = 0.0;
      SpectralEstimate est(rms, cutoff, minsig);
      est.setQ(q);
      const SpectralList &slist = est.estimate(xy, y);
      r = slist.nelements();
      cout << "Found (using rms, cutoff, minsig, q): (" <<
	rms << ", " << cutoff << ", " << minsig << ", " << q << ") " <<
	r << " estimates" << endl;
      for (Int i=0; i<r; i++) {
	cout << "Estimate " << i << ": " << *slist[i] << endl;
      };
    }
    {
      q = 2;
      rms = cutoff = minsig = 0.0;
      SpectralEstimate est(rms, cutoff, minsig);
      est.setQ(q);
      const SpectralList &slist = est.estimate(xy, y);
      r = slist.nelements();
      cout << "Found (using rms, cutoff, minsig, q): (" <<
	rms << ", " << cutoff << ", " << minsig << ", " << q << ") " <<
	r << " estimates" << endl;
      for (Int i=0; i<r; i++) {
	cout << "Estimate " << i << ": " << *slist[i] << endl;
      };
    }
  }

  // test fitter
  {
    // Number of components
    const uInt ncomp = 4;
    // Component data
    Double ampl[ncomp] = { 1, 2.5, 0.5, 5 };
    Double center[ncomp] = { 0.25, 0.5, 0.75, 0.3 };
    Double sigma[ncomp] = { 2, 4, 8, 3 };
    MLCG genit;
    Normal noise(&genit, 0.0, 0.1);
    Vector<Double> freq(1024);
    Vector<Float> ffreq(1024);
    for (uInt i=0; i<1024; i++) {
      freq(i) = 1400 + i*10.0/1024.0;
      ffreq(i) = freq(i);
    };

    try {
      cout << "Test SpectralFit" << endl;
      cout << "---------------------------------------------------" << endl;
      
      GaussianSpectralElement el[ncomp] = {
	GaussianSpectralElement(ampl[0],
			center[0]*10.+1400., sigma[0]*10./1024.),
	GaussianSpectralElement(ampl[1],
			center[1]*10.+1400., sigma[1]*10./1024.),
	GaussianSpectralElement(ampl[2],
			center[2]*10.+1400., sigma[2]*10./1024.),
	GaussianSpectralElement(ampl[3],
			center[3]*10.+1400., sigma[3]*10./1024.) };
      
      cout << "Spectral elements: " << endl;
      for (uInt j=0; j<ncomp; j++) cout << el[j] << endl;
      cout << "---------------------------------------------------" << endl;
      
      Vector<Double> dat(1024);
      Vector<Float> fdat(1024);
      for (uInt i=0; i<1024; i++) {
	dat(i) = 0;
	for (uInt j=0; j<ncomp; j++) {
	  dat(i) += el[j].getAmpl()*exp(-(freq(i)-el[j].getCenter())*
					(freq(i)-el[j].getCenter())*log(16.0)/
					el[j].getSigma()/
					el[j].getSigma());
	};
	dat(i) += noise();
	fdat(i) = dat(i);
      };
      cout << "Data for frequencies " << freq(300) << " - " << freq(315) <<
	endl;
      for (uInt i=300; i<316; i++) cout << freq(i) << ": " << dat(i) << endl;
      cout << "---------------------------------------------------" << endl;
      
      cout << "Specify fitter on the 4 gaussians:" << endl;
      SpectralFit fitter;
      for (uInt i=0; i<ncomp; i++) fitter.addFitElement(el[i]);
      for (uInt i=0; i<fitter.list().nelements(); i++) {
	cout << *fitter.list()[i] << endl;
      };

      cout << "---------------------------------------------------" << endl;
      cout << "Execute the fitter, can do (1:Y): " <<
	fitter.fit(dat,freq) << endl;
      cout << "The results: " << endl;
      Vector<Double> tmp;
      for (uInt i=0; i<fitter.list().nelements(); i++) {
	cout << *fitter.list()[i];
	fitter.list()[i]->get(tmp);
	cout << "Parameters: " << tmp << endl;
	fitter.list()[i]->getError(tmp);
	cout << "Errors:     " << tmp << endl << endl;
      };
      cout << ">>>" << endl;
      cout << "Number of iterations: " << fitter.nIterations() << endl;
      cout << "<<<" << endl;
      cout << "---------------------------------------------------" << endl;
      
      {
	el[3].fixAmpl();
	SpectralFit fitter;
	for (uInt i=0; i<ncomp; i++) fitter.addFitElement(el[i]);
	for (uInt i=0; i<fitter.list().nelements(); i++) {
	  cout << *fitter.list()[i] << endl;
	};
	
	cout << "---------------------------------------------------" << endl;
	cout << "Execute the fitter again with a fixed ampl, can do (1:Y): " <<
	  fitter.fit(dat,freq) << endl;
	cout << "The results: " << endl;
	Vector<Double> tmp;
	for (uInt i=0; i<fitter.list().nelements(); i++) {
	  cout << *fitter.list()[i];
	  fitter.list()[i]->get(tmp);
	  cout << "Parameters: " << tmp << endl;
	  fitter.list()[i]->getError(tmp);
	  cout << "Errors:     " << tmp << endl << endl;
	};
	cout << ">>>" << endl;
	cout << "Number of iterations: " << fitter.nIterations() << endl;
	cout << "<<<" << endl;
	el[3].fixAmpl(False);
      }
      cout << "---------------------------------------------------" << endl;
      
      cout << "Different start values: " << endl;
      for (uInt i=0; i<ncomp; i++) {
	el[i].setAmpl(0.9*el[i].getAmpl());
	el[i].setSigma(0.95*el[i].getSigma());
	fitter.setFitElement(i, el[i]);
      };
      for (uInt i=0; i<fitter.list().nelements(); i++) {
	cout << *fitter.list()[i] << endl;
      };
      cout << "Execute the fitter, can do (1:Y): " <<
	fitter.fit(dat, freq) << endl;
      cout << "The results: " << endl;
      for (uInt i=0; i<fitter.list().nelements(); i++) {
	cout << *fitter.list()[i] << endl;
      };
      cout << ">>>" << endl;
      cout << "Number of iterations: " << fitter.nIterations() << endl;
      cout << "<<<" << endl;
      
      cout << "---------------------------------------------------" << endl;
      cout << "Differences: " << endl;
      Vector<Double> xdat(1024);
      Vector<Float> fxdat(1024);
      Double mx(-1e6);
      Double mn(1e6);
      Double avg(0);
      Double sg(0);
      xdat = dat;
      fitter.list().residual(xdat, freq);
      convertArray(fxdat, xdat);
      mx = max(xdat);
      mn = min(xdat);
      avg = sum(xdat);
      sg = sum(xdat*xdat);
      avg /= 1024.;
      sg = sqrt(sg/1024./1023.);
      cout << "Min difference: " << mn <<
	", max: " << mx << ", average: " << avg <<
	", sigma: " << sg << endl;
      cout << "---------------------------------------------------" << endl;
      
      cout << "---------------------------------------------------" << endl;
      cout << "Estimates: " << endl;
      SpectralEstimate mest(0.1, 0.5);
      mest.setQ(5);
      const SpectralList &slist = mest.estimate(fdat);
      Int mr = slist.nelements();
      cout << "Found " << mr << " estimates" << endl;
      for (Int i=0; i<mr; i++) {
	cout << "Estimate " << i << ": " << *slist[i] << endl;
      };
      fxdat= fdat;
      slist.residual(fxdat);
      for (Int j=0; j<mr; j++) {
	slist.evaluate(fxdat);
      };
      
      cout << "---------------------------------------------------" << endl;
      cout << "Limited number of estimates: " << endl;
      {
	SpectralEstimate mest(0.1, 0.5, 0.0, 4);
	mest.setQ(5);
	const SpectralList &slist = mest.estimate(fdat);
	Int mr = slist.nelements();
	cout << "Found " << mr << " estimates" << endl;
	for (Int i=0; i<mr; i++) {
	  cout << "Estimate " << i << ": " << *slist[i] << endl;
	};
	fxdat= fdat;
	slist.residual(fxdat);
	for (Int j=0; j<mr; j++) {
	  slist.evaluate(fxdat);
	};
      }      
      cout << "---------------------------------------------------" << endl;
      
    } catch (AipsError x) {
      cout << x.getMesg() << endl;
      return 1;
    };
  }
  
  return 0;
}
