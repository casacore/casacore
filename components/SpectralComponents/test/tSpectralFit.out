Test RecordInterface
---------------------------------------------------
toRecord
fromRecord

Test SpectralEstimate
---------------------------------------------------
Made spectrum with components: 
GAUSSIAN element: 
  Amplitude: 5, 0
  Center:    45, 0
  Sigma:     1.1, 0

GAUSSIAN element: 
  Amplitude: 4, 0
  Center:    50, 0
  Sigma:     1.1, 0

GAUSSIAN element: 
  Amplitude: 6, 0
  Center:    55, 0
  Sigma:     1.1, 0

Found (using rms, cutoff, minsig, q): (0.5, 1, 0.5, 2) 3 estimates
Estimate 0: GAUSSIAN element: 
  Amplitude: 6.00013, 0
  Center:    55.0715, 0
  Sigma:     1.33398, 0

Estimate 1: GAUSSIAN element: 
  Amplitude: 5.00013, 0
  Center:    44.9152, 0
  Sigma:     1.32309, 0

Estimate 2: GAUSSIAN element: 
  Amplitude: 4.00036, 0
  Center:    50.0043, 0
  Sigma:     1.26335, 0

Minimum, Maximum residuals: -0.961305, 0.0086807
Window (using rms, cutoff, minsig, q): (0.5, 1, 0.5, 2) 0 estimates
Found (using rms, cutoff, minsig, q): (0.5, 1, 0.5, 1) 3 estimates
Estimate 0: GAUSSIAN element: 
  Amplitude: 6.00013, 0
  Center:    54.9829, 0
  Sigma:     1.17129, 0

Estimate 1: GAUSSIAN element: 
  Amplitude: 5.00013, 0
  Center:    45.0202, 0
  Sigma:     1.17355, 0

Estimate 2: GAUSSIAN element: 
  Amplitude: 4.00036, 0
  Center:    50.0182, 0
  Sigma:     1.38534, 0

Found (using rms, cutoff, minsig, q): (0, 0, 0, 1) 3 estimates
Estimate 0: GAUSSIAN element: 
  Amplitude: 6.00013, 0
  Center:    54.9829, 0
  Sigma:     1.17129, 0

Estimate 1: GAUSSIAN element: 
  Amplitude: 5.00013, 0
  Center:    45.0202, 0
  Sigma:     1.17355, 0

Estimate 2: GAUSSIAN element: 
  Amplitude: 4.00036, 0
  Center:    50.0182, 0
  Sigma:     1.38534, 0

Found (using rms, cutoff, minsig, q): (0, 0, 0, 2) 3 estimates
Estimate 0: GAUSSIAN element: 
  Amplitude: 6.00013, 0
  Center:    55.0715, 0
  Sigma:     1.33398, 0

Estimate 1: GAUSSIAN element: 
  Amplitude: 5.00013, 0
  Center:    44.9152, 0
  Sigma:     1.32309, 0

Estimate 2: GAUSSIAN element: 
  Amplitude: 4.00036, 0
  Center:    50.0043, 0
  Sigma:     1.26335, 0

Test SpectralFit
---------------------------------------------------
Spectral elements: 
GAUSSIAN element: 
  Amplitude: 1, 0
  Center:    1402.5, 0
  Sigma:     0.0195312, 0

GAUSSIAN element: 
  Amplitude: 2.5, 0
  Center:    1405, 0
  Sigma:     0.0390625, 0

GAUSSIAN element: 
  Amplitude: 0.5, 0
  Center:    1407.5, 0
  Sigma:     0.078125, 0

GAUSSIAN element: 
  Amplitude: 5, 0
  Center:    1403, 0
  Sigma:     0.0292969, 0

---------------------------------------------------
Data for frequencies 1402.93 - 1403.08
1402.93: 0.219133
1402.94: -0.480645
1402.95: -0.281389
1402.96: 0.0423511
1402.97: 0.0837382
1402.98: 1.05514
1402.99: 3.14053
1403: 4.93017
1403.01: 4.53475
1403.02: 1.54477
1403.03: 0.247699
1403.04: -0.50962
1403.05: 0.0446798
1403.06: -0.0518962
1403.07: -0.750935
1403.08: 0.499586
---------------------------------------------------
Specify fitter on the 4 gaussians:
GAUSSIAN element: 
  Amplitude: 1, 0
  Center:    1402.5, 0
  Sigma:     0.0195312, 0

GAUSSIAN element: 
  Amplitude: 2.5, 0
  Center:    1405, 0
  Sigma:     0.0390625, 0

GAUSSIAN element: 
  Amplitude: 0.5, 0
  Center:    1407.5, 0
  Sigma:     0.078125, 0

GAUSSIAN element: 
  Amplitude: 5, 0
  Center:    1403, 0
  Sigma:     0.0292969, 0

---------------------------------------------------
Execute the fitter, can do (1:Y): 1
The results: 
GAUSSIAN element: 
  Amplitude: 0.917797, 0.310674
  Center:    1402.5, 0.00347673
  Sigma:     0.00882989, 0.00343021
Parameters: [0.917797, 1402.5, 0.00882989]
Errors:     [0.310674, 0.00347673, 0.00343021]

GAUSSIAN element: 
  Amplitude: 2.76117, 0.238196
  Center:    1405, 0.00151172
  Sigma:     0.0151761, 0.00151172
Parameters: [2.76117, 1405, 0.0151761]
Errors:     [0.238196, 0.00151172, 0.00151172]

GAUSSIAN element: 
  Amplitude: 0.612907, 0.18968
  Center:    1407.48, 0.0085527
  Sigma:     0.0239303, 0.0085527
Parameters: [0.612907, 1407.48, 0.0239303]
Errors:     [0.18968, 0.0085527, 0.0085527]

GAUSSIAN element: 
  Amplitude: 5.24752, 0.272235
  Center:    1403, 0.000695977
  Sigma:     0.0116179, 0.000695945
Parameters: [5.24752, 1403, 0.0116179]
Errors:     [0.272235, 0.000695977, 0.000695945]

>>>
Number of iterations: 29
<<<
---------------------------------------------------
GAUSSIAN element: 
  Amplitude: 1, 0
  Center:    1402.5, 0
  Sigma:     0.0195312, 0

GAUSSIAN element: 
  Amplitude: 2.5, 0
  Center:    1405, 0
  Sigma:     0.0390625, 0

GAUSSIAN element: 
  Amplitude: 0.5, 0
  Center:    1407.5, 0
  Sigma:     0.078125, 0

GAUSSIAN element: 
  Amplitude: 5, 0 (fixed)
  Center:    1403, 0
  Sigma:     0.0292969, 0

---------------------------------------------------
Execute the fitter again with a fixed ampl, can do (1:Y): 1
The results: 
GAUSSIAN element: 
  Amplitude: 0.917772, 0.310662
  Center:    1402.5, 0.00347618
  Sigma:     0.00883038, 0.00342961
Parameters: [0.917772, 1402.5, 0.00883038]
Errors:     [0.310662, 0.00347618, 0.00342961]

GAUSSIAN element: 
  Amplitude: 2.76117, 0.23818
  Center:    1405, 0.00151162
  Sigma:     0.0151761, 0.00151162
Parameters: [2.76117, 1405, 0.0151761]
Errors:     [0.23818, 0.00151162, 0.00151162]

GAUSSIAN element: 
  Amplitude: 0.61293, 0.18968
  Center:    1407.48, 0.00855088
  Sigma:     0.0239286, 0.00855088
Parameters: [0.61293, 1407.48, 0.0239286]
Errors:     [0.18968, 0.00855088, 0.00855088]

GAUSSIAN element: 
  Amplitude: 5, 0 (fixed)
  Center:    1403, 0.000741387
  Sigma:     0.0119707, 0.00060533
Parameters: [5, 1403, 0.0119707]
Errors:     [0, 0.000741387, 0.00060533]

>>>
Number of iterations: 26
<<<
---------------------------------------------------
Different start values: 
GAUSSIAN element: 
  Amplitude: 0.9, 0
  Center:    1402.5, 0
  Sigma:     0.0185547, 0

GAUSSIAN element: 
  Amplitude: 2.25, 0
  Center:    1405, 0
  Sigma:     0.0371094, 0

GAUSSIAN element: 
  Amplitude: 0.45, 0
  Center:    1407.5, 0
  Sigma:     0.0742188, 0

GAUSSIAN element: 
  Amplitude: 4.5, 0
  Center:    1403, 0
  Sigma:     0.027832, 0

Execute the fitter, can do (1:Y): 1
The results: 
GAUSSIAN element: 
  Amplitude: 0.917785, 0.310679
  Center:    1402.5, 0.00347653
  Sigma:     0.00883013, 0.00342999

GAUSSIAN element: 
  Amplitude: 2.76117, 0.238196
  Center:    1405, 0.00151172
  Sigma:     0.0151761, 0.00151172

GAUSSIAN element: 
  Amplitude: 0.61293, 0.189693
  Center:    1407.48, 0.00855143
  Sigma:     0.0239285, 0.00855143

GAUSSIAN element: 
  Amplitude: 5.24752, 0.272235
  Center:    1403, 0.000695977
  Sigma:     0.0116179, 0.000695945

>>>
Number of iterations: 32
<<<
---------------------------------------------------
Differences: 
Min difference: -1.09994, max: 1.01271, average: 0.00978336, sigma: 0.0100325
---------------------------------------------------
---------------------------------------------------
Estimates: 
Found 20 estimates
Estimate 0: GAUSSIAN element: 
  Amplitude: 4.93017, 0
  Center:    307.195, 0
  Sigma:     3.05174, 0

Estimate 1: GAUSSIAN element: 
  Amplitude: 2.79525, 0
  Center:    511.819, 0
  Sigma:     3.29148, 0

Estimate 2: GAUSSIAN element: 
  Amplitude: 0.906691, 0
  Center:    766.653, 0
  Sigma:     2.31423, 0

Estimate 3: GAUSSIAN element: 
  Amplitude: 0.90563, 0
  Center:    501.372, 0
  Sigma:     3.60214, 0

Estimate 4: GAUSSIAN element: 
  Amplitude: 0.861827, 0
  Center:    839.949, 0
  Sigma:     3.43708, 0

Estimate 5: GAUSSIAN element: 
  Amplitude: 0.846625, 0
  Center:    112.878, 0
  Sigma:     2.28534, 0

Estimate 6: GAUSSIAN element: 
  Amplitude: 0.751603, 0
  Center:    871.157, 0
  Sigma:     2.78939, 0

Estimate 7: GAUSSIAN element: 
  Amplitude: 0.699624, 0
  Center:    36.4464, 0
  Sigma:     3.65743, 0

Estimate 8: GAUSSIAN element: 
  Amplitude: 0.682235, 0
  Center:    395.917, 0
  Sigma:     3.5893, 0

Estimate 9: GAUSSIAN element: 
  Amplitude: 0.663379, 0
  Center:    450.12, 0
  Sigma:     0.548263, 0

Estimate 10: GAUSSIAN element: 
  Amplitude: 0.640783, 0
  Center:    827.254, 0
  Sigma:     3.35359, 0

Estimate 11: GAUSSIAN element: 
  Amplitude: 0.616448, 0
  Center:    245.555, 0
  Sigma:     3.28022, 0

Estimate 12: GAUSSIAN element: 
  Amplitude: 0.578568, 0
  Center:    120.043, 0
  Sigma:     1.45528, 0

Estimate 13: GAUSSIAN element: 
  Amplitude: 0.569486, 0
  Center:    663.971, 0
  Sigma:     1.81295, 0

Estimate 14: GAUSSIAN element: 
  Amplitude: 0.56377, 0
  Center:    611.848, 0
  Sigma:     2.99822, 0

Estimate 15: GAUSSIAN element: 
  Amplitude: 0.561451, 0
  Center:    937.91, 0
  Sigma:     2.32834, 0

Estimate 16: GAUSSIAN element: 
  Amplitude: 0.54666, 0
  Center:    336.559, 0
  Sigma:     1.8761, 0

Estimate 17: GAUSSIAN element: 
  Amplitude: 0.524891, 0
  Center:    443.644, 0
  Sigma:     0.809077, 0

Estimate 18: GAUSSIAN element: 
  Amplitude: 0.523377, 0
  Center:    645.266, 0
  Sigma:     3.01063, 0

Estimate 19: GAUSSIAN element: 
  Amplitude: 0.515985, 0
  Center:    446.61, 0
  Sigma:     1.18317, 0

---------------------------------------------------
Limited number of estimates: 
Found 4 estimates
Estimate 0: GAUSSIAN element: 
  Amplitude: 4.93017, 0
  Center:    307.195, 0
  Sigma:     3.05174, 0

Estimate 1: GAUSSIAN element: 
  Amplitude: 2.79525, 0
  Center:    511.819, 0
  Sigma:     3.29148, 0

Estimate 2: GAUSSIAN element: 
  Amplitude: 0.906691, 0
  Center:    766.653, 0
  Sigma:     2.31423, 0

Estimate 3: GAUSSIAN element: 
  Amplitude: 0.90563, 0
  Center:    501.372, 0
  Sigma:     3.60214, 0

---------------------------------------------------
